### DISCUSSION TITLE
Maven Java Agent Config not working

### DISCUSSION THREAD

**Discussion by GerMoranBYON** [mockito/mockito#3524] [Q&A]
The following is described as away to provide the agent via config through maven for sure-fire.

However the documentation is mainly centralised around Gradle, 

https://github.com/mockito/mockito/blob/ebec7f22db77f92af87386ce229588cac93e200c/mockito-core/src/main/java/org/mockito/Mockito.java#L224C1-L224C96 

I am unable to add the properties.

**Comment by RockyMM** [comment]
Not an answer, I am facing the same issue. @GerMoranBYON , have you find a solution?

**Comment by RockyMM** [comment]
Actually, I found it on this [SO](https://stackoverflow.com/a/50704624/339219).

We need to add the maven-dependency-plugin with `properties` execution:

```xml
            <plugin>
                <artifactId>maven-dependency-plugin</artifactId>
                <version>3.8.1</version>
                <executions>
                    <execution>
                        <goals>
                            <goal>properties</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>
```

Also, there are some hints how to make it work in IntelliJ IDEA and Eclipse, but this was enough for me to make it work.

**Reply by leluna** [reply]
Link to the Mockito Documentation: https://javadoc.io/static/org.mockito/mockito-core/5.23.0/org.mockito/org/mockito/Mockito.html#0.3 with an example for Maven

**Comment by yuvkun10** [comment]
Here's the complete Maven configuration to attach the Mockito agent via Surefire, building on @RockyMM's answer:

The `maven-dependency-plugin` with the `properties` goal sets a property `org.mockito:mockito-core` to the path of the JAR. You then reference that in Surefire's `argLine`.

**Full working configuration:**

```xml
<build>
  <plugins>
    <!-- Step 1: Resolve the mockito-core JAR path into a Maven property -->
    <plugin>
      <artifactId>maven-dependency-plugin</artifactId>
      <version>3.8.1</version>
      <executions>
        <execution>
          <goals>
            <goal>properties</goal>
          </goals>
        </execution>
      </executions>
    </plugin>

    <!-- Step 2: Pass it as a javaagent to Surefire -->
    <plugin>
      <artifactId>maven-surefire-plugin</artifactId>
      <version>3.3.0</version>
      <configuration>
        <argLine>-javaagent:${org.mockito:mockito-core:jar}</argLine>
      </configuration>
    </plugin>
  </plugins>
</build>
```

The property name format is `groupId:artifactId:type` — so `${org.mockito:mockito-core:jar}` expands to the absolute path of mockito-core's JAR in your local repository.

**Why the dependency-plugin `properties` goal is needed:**

Without it, Surefire doesn't know the on-disk path of the JAR. The `properties` goal runs during `initialize` phase and sets Maven properties like `org.mockito:mockito-core:jar=/home/user/.m2/.../mockito-core-5.x.x.jar`.

**For IntelliJ IDEA:** Add a VM option in Run → Edit Configurations → VM options:
```
-javaagent:/home/user/.m2/repository/org/mockito/mockito-core/5.11.0/mockito-core-5.11.0.jar
```

Or use the official [Mockito Maven Plugin](https://github.com/mockito/mockito-maven-plugin) which handles this automatically.

Hope this helps! Let me know if you need clarification.
